### DISCUSSION TITLE
DistroHopper

### DISCUSSION THREAD

**Discussion by zen0bit** [sxiii/linux-downloader#3] [Show and tell]
I am creator of [DistroHopper](https://osowoso.github.io/DistroHopper/)

![hop](https://user-images.githubusercontent.com/6384793/226753996-e181460e-9879-498c-b0ba-9328161fd239.svg)

Looking for way to automatize downloading ISOs.

And found your script

I am actually using quickget from [quickemu](https://github.com/quickemu-project/quickemu) to get ISOs.

Your script support more OSes

quickget possible support all editions and releases not only newest ones.

I am looking for way to combine your and quickget...

Let me know if you are interested...

**Comment by sxiii** [comment]
Greetingd @zen0bit ! Thanks for your message!

I would be super-happy to join your efforts. Basically, I was thinking in the long run to develop something similar to your tool in the future. Cool that you're having this project!

About support for multiple releases, this was planned and somewhat it's included in the logic of the script - I just didn't bothered to add more different releases... Also maybe worth thinking about my script project structure again.

I haven't seen quickget script. Thanks for the info! The downside to these scripts is, of course, the fact, that distributions tend to move things around from time to time. 

I'm seeing two solutions right now to this:
A) Intermediary solution (fast & easy to make but a bit dirty). Run a curl or similar script every hour (or every day) against list of several hundreds distro's (ISO's resulting endpoints) to see how likely that the ISO is still there. If there's some problems, record the error and send alert (aka 'monitoring'). So this would need some sort of a backend I guess. I had looked into https://pkgs.org, ArchLinux mirrorlist status, and some Fedora & AltLinux automations a lot and I think that would be cool to have something similar, but for ISOs/IMGs instead of packages/mirrors.
B) Full 'proper' solution. Build a 'protocol' or a 'own standard' to ship ISOs from different vendors. This could be done once and solve the problem forever, but it would require *some* support from Distributions themselves. As an example, it could be a text file which is located at the root of the mirror. Say, releases.md or releases.csv. Then, it could include something like
```
Distro,Version,Architecture,URLs
Fedora,37,amd64,https://download.getfedora.org/fedora37-amd64.iso,https://backupmirror.fedorafriends.com/f37-amd64.iso
Ubuntu,23.04,amd64,https://mirror.ubuntu.com/ubuntu-2304-amd64.iso,https://getubuntu.com/ubuntu-2304-amd64.torrent
Arch,Rolling,amd64,https://download.archlinux.org/arch.torrent
```
Which we could parse automatically with relative ease, then the distro makers could add a CI/CD procedure to their build toolchain and they would have automatically updated release file. It's both for their and our convenience (no need to hack the way around downloads, awk/grep etc). Also, we could add checksums (right after the ISOs maybe) and include automated checksum control.

Of course, for this to happen, first 2-3-4 distros should accommodate this 'protocol', so we'll have to develop it, publish it, and showcase it. But it's an interesting idea IMO, reminds me a little about 'loopback' file in the ISO which allows to find and load ISOs 'live' by Grub itself (this loopback.cfg file was added by plenty of distros, but not all).

Either way, I'm eager to listen to your thoughts and feedback on my project and ways of going forward!

Thanks :)
